services/components/containers/bot_client_test.go

```diff
@@ -97,7 +105,55 @@ func (s *BotClientTestSuite) TestLaunchBot_Exists() {
 		s.client.EXPECT().AttachNetwork(gomock.Any(), testContainerID, testBotNetworkID).Return(nil)
 	}
 
+	resources := &docker.ContainerResources{
+		CPUStats: docker.CPUStats{
+			CPUUsage: docker.CPUUsage{
+				TotalUsage: 33,
+			},
+		},
+		MemoryStats: docker.MemoryStats{
+			Usage: 100,
+		},
+		NetworkStats: map[string]docker.NetworkStats{
+			"eth0": {
+				RxBytes: 123,
+				TxBytes: 456,
+			},
+		},
+	}
+	s.client.EXPECT().ContainerStats(gomock.Any(), botConfig.ContainerName()).Return(resources, nil).Times(1)
+	s.msgClient.EXPECT().PublishProto(messaging.SubjectMetricAgent, gomock.Any()).Do(func(v1, v2 interface{}) {
+		metrics := v2.(*protocol.AgentMetricList)
+		assert.Len(s.T(), metrics.Metrics, 4)
+
+		// CPU metric
+		assert.Equal(s.T(), botConfig.ID, metrics.Metrics[0].AgentId)
+		assert.Equal(s.T(), domain.MetricDockerResourcesCPU, metrics.Metrics[0].Name)
+		assert.Equal(s.T(), float64(33), metrics.Metrics[0].Value)
+
+		// Memory metric
+		assert.Equal(s.T(), botConfig.ID, metrics.Metrics[1].AgentId)
+		assert.Equal(s.T(), domain.MetricDockerResourcesMemory, metrics.Metrics[1].Name)
+		assert.Equal(s.T(), float64(100), metrics.Metrics[1].Value)
+
+		// Network bytes received metric
+		assert.Equal(s.T(), botConfig.ID, metrics.Metrics[3].AgentId)
+		assert.Equal(s.T(), domain.MetricDockerResourcesNetworkReceive, metrics.Metrics[3].Name)
+		assert.Equal(s.T(), float64(123), metrics.Metrics[3].Value)
+
+		// Network bytes sent metric
+		assert.Equal(s.T(), botConfig.ID, metrics.Metrics[2].AgentId)
+		assert.Equal(s.T(), domain.MetricDockerResourcesNetworkSent, metrics.Metrics[2].Name)
+		assert.Equal(s.T(), float64(456), metrics.Metrics[2].Value)
+	})
+
 	s.r.NoError(s.botClient.LaunchBot(context.Background(), botConfig))
+
+	ticker := initTicker(DockerResourcesPollingInterval)
+	defer ticker.Stop()
+
+	<-ticker.C
+	time.Sleep(100 * time.Millisecond)
```

Comment:
Why do we need to wait here?

Reply:
I think it might be better to define a channel before `PublishProto().Do()`, listen to that here and close in `Do()` arg func, what do you think? Then we can actually exit the test when all expected calls are executed. Using `time.Sleep()` like this can cause flaky tests sometimes.

Also, I think `LaunchBot()` is already calling `initTicker()` at some point. Is there any other reason why we are redoing that in the unit test?